community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/InvocationTracer.java

```diff
@@ -0,0 +1,282 @@
+import org.neo4j.kernel.api.index.IndexDescriptor;
+
+import static java.lang.String.format;
+
+public class InvocationTracer<C> implements InvocationHandler
+{
+    private static final String LINE_SEPARATOR = System.getProperty( "line.separator" );
+
+    private final String generatedClassPackage;
+    private final String generatedClassName;
+
+    private final StringBuilder builder;
+    private final Class<C> interfaceClass;
+
+    public InvocationTracer( String generatedClassNameWithPackage, Class<C> interfaceClass )
+    {
+        if ( generatedClassNameWithPackage.contains( "%" ) )
+        {
+            throw new IllegalArgumentException( "Format character in generated class name: " + generatedClassNameWithPackage );
+        }
+
+        Pair<String, String> parsedGenerated = parseQualifiedClassName( generatedClassNameWithPackage );
+        this.generatedClassPackage = parsedGenerated.first();
+        this.generatedClassName = parsedGenerated.other();
+
+        this.interfaceClass = interfaceClass;
+
+        this.builder = new StringBuilder();
+    }
+
+    public C newProxy()
+    {
+        return newProxy( interfaceClass );
+    }
+
+    public <P extends C> P newProxy( Class<P> proxyClass )
+    {
+        ClassLoader classLoader = proxyClass.getClassLoader();
+        return proxyClass.cast(
+            Proxy.newProxyInstance( classLoader, new Class[] { proxyClass }, this )
+        );
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder result = new StringBuilder();
+        printPreamble( result );
+        printTrace( result );
+        printAppendix( result );
+        return result.toString();
+    }
+
+    public String getGeneratedPackageName()
+    {
+        return generatedClassPackage;
+    }
+
+    public String getGeneratedClassName()
+    {
+        return generatedClassName;
+    }
+
+    private void printPreamble( StringBuilder builder )
+    {
+        String interfaceClassName = interfaceClass.getCanonicalName();
+        println( builder, "package %s;", generatedClassPackage );
+        println( builder );
+        println( builder, "import org.neo4j.kernel.api.constraints.UniquenessConstraint;");
+        println( builder, "import org.neo4j.kernel.api.index.IndexDescriptor;");
+        println( builder, "import org.neo4j.kernel.impl.util.dbstructure.Visitable;");
+        println( builder );
+        println( builder, "//" );
+        println( builder, "// GENERATED FILE. DO NOT EDIT. " );
+        println( builder, "//" );
+        println( builder, "// This has been generated by %s", getClass().getCanonicalName() );
+        println( builder, "//" );
+        println( builder );
+        println( builder, "public class %s implements Visitable<%s>", generatedClassName, interfaceClassName );
+        println( builder, "{" );
+        println( builder, "    public final static Visitable<%s> INSTANCE = new %s();", interfaceClassName, generatedClassName );
```

Comment:
Hmm I was suspecting that wouldn't work with mocking but even that can be solved it seems (http://stackoverflow.com/questions/15939023/how-to-mock-an-enum-singleton-class-using-mockito-powermock). So, yeah that's a good change then.